src/librbd/Utils.cc

```diff
@@ -339,7 +340,7 @@ int init_rados(CephContext *cct,
     }
   }
 
-  if (!g_ceph_context->_conf->admin_socket.empty()) {
+  if (set_global_admin_socket && !g_ceph_context->_conf->admin_socket.empty()) {
```

Comment:
Actually, I am not sure about this `set_global_admin_socket`. I don't see any reference to `g_ceph_context` in librbd code, and I don't think we want to introduce it here.

I would either pass `const std::srting &admin_socket` instead of `bool set_global_admin_socke` and set it if it is not empty, or just entirely moved this code out of this function and did it explicitly in rbd-mirror after calling `init_rados`.
And if you decide to leave the code here, then I would change `set_val_or_die` to just `set_val` and returned with the error if it failed (similarly to what we do below for "mon_host").

Reply:
So my preference is to set the admin socket in rbd_mirror, outside of `init_rados`.
However I have no clue on what it does, and if it's ok to set it after `rados->connect()` was called.
Can you vouch that it will not change anything for rbd_mirror?

Reply:
Ah, indeed, good point. I think we need to properly set the socket file name before initializing the new context, so the admin socket is created using the new name. So then yes, it leaves us with the only option: just provide the socket name as a parameter instead of trying to get it from the global context.

Reply:
Sorry, I was too fast with my suggestion. I thought `g_ceph_context->_conf->admin_socket` used for the new socket name, but actually it does not. We only need it to check if it is not empty. So a new suggestion :-)

Just change the condition to `if (set_global_admin_socket)` here and check for `g_ceph_context->_conf->admin_socket.empty()` in the rbd-mirro code when calculating `set_global_admin_socket` param?

Reply:
Right, I also did not notice that it is not used for the new socket.
So perhaps it's safe to move it all out to rbd_mirror code?
Anyway, if your goal is to eliminate the use of `g_ceph_context` from `init_rados`, we need to move the setting of the admin socket configuration as well (not just the check).

Reply:
Or actually it still could be `const std::srting &admin_socket` instead of `bool set_global_admin_socket`, you just need to set it to `"$run_dir/$name.$pid.$cluster.$cctid.asok"` in the rbd-mirror code (or leave empty if the global socket name is empty). I think I would like this more as a more generic, but I think I will fine with any variant that does not require `g_ceph_context` usage here.

Reply:
> Anyway, if your goal is to eliminate the use of g_ceph_context from init_rados, we need to move the setting of the admin socket configuration as well (not just the check).

Note, `g_ceph_context` is used only to see if the socket name is empty or not in the main (global) configuration. It is not used for setting the admin socket here (it is set in the new context).

An empty value of "admin_socket" configuration parameter just has a special meaning: do not start/initialize the admin socket at all. That is why we check if it is empty or not before setting the socket name for the new context. We can't just reuse the global admin socket name in the new context, because it is already used and we would have "socket exist" error when trying to initialize the admin socket for the new context. That is why the rbd-mirror change the name to something unique (that includes `$cctid`).

Reply:
My bad. This is why I have to re-check the code before commenting :)
Thanks!